File: src/services/sortModes.ts
import { invokeTauriOrNull } from "./tauriInvoke";
import type { CliKey } from "./providers";

export type SortModeSummary = {
  id: number;
  name: string;
  created_at: number;
  updated_at: number;
};

export type SortModeActiveRow = {
  cli_key: CliKey;
  mode_id: number | null;
  updated_at: number;
};

export async function sortModesList() {
  return invokeTauriOrNull<SortModeSummary[]>("sort_modes_list");
}

export async function sortModeCreate(input: { name: string }) {
  return invokeTauriOrNull<SortModeSummary>("sort_mode_create", {
    name: input.name,
  });
}

export async function sortModeRename(input: { mode_id: number; name: string }) {
  return invokeTauriOrNull<SortModeSummary>("sort_mode_rename", {
    modeId: input.mode_id,
    name: input.name,
  });
}

export async function sortModeDelete(input: { mode_id: number }) {
  return invokeTauriOrNull<boolean>("sort_mode_delete", {
    modeId: input.mode_id,
  });
}

export async function sortModeActiveList() {
  return invokeTauriOrNull<SortModeActiveRow[]>("sort_mode_active_list");
}

export async function sortModeActiveSet(input: { cli_key: CliKey; mode_id: number | null }) {
  return invokeTauriOrNull<SortModeActiveRow>("sort_mode_active_set", {
    cliKey: input.cli_key,
    modeId: input.mode_id,
  });
}

export async function sortModeProvidersList(input: { mode_id: number; cli_key: CliKey }) {
  return invokeTauriOrNull<number[]>("sort_mode_providers_list", {
    modeId: input.mode_id,
    cliKey: input.cli_key,
  });
}

export async function sortModeProvidersSetOrder(input: {
  mode_id: number;
  cli_key: CliKey;
  ordered_provider_ids: number[];
}) {
  return invokeTauriOrNull<number[]>("sort_mode_providers_set_order", {
    modeId: input.mode_id,
    cliKey: input.cli_key,
    orderedProviderIds: input.ordered_provider_ids,
  });
}

File: src-tauri/src/commands/sort_modes.rs
//! Usage: Provider sort modes related Tauri commands.

use crate::app_state::{ensure_db_ready, DbInitState};
use crate::{blocking, sort_modes};

#[tauri::command]
pub(crate) async fn sort_modes_list(
    app: tauri::AppHandle,
    db_state: tauri::State<'_, DbInitState>,
) -> Result<Vec<sort_modes::SortModeSummary>, String> {
    ensure_db_ready(app.clone(), db_state.inner()).await?;
    blocking::run("sort_modes_list", move || sort_modes::list_modes(&app)).await
}

#[tauri::command]
pub(crate) async fn sort_mode_create(
    app: tauri::AppHandle,
    db_state: tauri::State<'_, DbInitState>,
    name: String,
) -> Result<sort_modes::SortModeSummary, String> {
    ensure_db_ready(app.clone(), db_state.inner()).await?;
    blocking::run("sort_mode_create", move || {
        sort_modes::create_mode(&app, &name)
    })
    .await
}

#[tauri::command]
pub(crate) async fn sort_mode_rename(
    app: tauri::AppHandle,
    db_state: tauri::State<'_, DbInitState>,
    mode_id: i64,
    name: String,
) -> Result<sort_modes::SortModeSummary, String> {
    ensure_db_ready(app.clone(), db_state.inner()).await?;
    blocking::run("sort_mode_rename", move || {
        sort_modes::rename_mode(&app, mode_id, &name)
    })
    .await
}

#[tauri::command]
pub(crate) async fn sort_mode_delete(
    app: tauri::AppHandle,
    db_state: tauri::State<'_, DbInitState>,
    mode_id: i64,
) -> Result<bool, String> {
    ensure_db_ready(app.clone(), db_state.inner()).await?;
    blocking::run("sort_mode_delete", move || {
        sort_modes::delete_mode(&app, mode_id)?;
        Ok(true)
    })
    .await
}

#[tauri::command]
pub(crate) async fn sort_mode_active_list(
    app: tauri::AppHandle,
    db_state: tauri::State<'_, DbInitState>,
) -> Result<Vec<sort_modes::SortModeActiveRow>, String> {
    ensure_db_ready(app.clone(), db_state.inner()).await?;
    blocking::run("sort_mode_active_list", move || {
        sort_modes::list_active(&app)
    })
    .await
}

#[tauri::command]
pub(crate) async fn sort_mode_active_set(
    app: tauri::AppHandle,
    db_state: tauri::State<'_, DbInitState>,
    cli_key: String,
    mode_id: Option<i64>,
) -> Result<sort_modes::SortModeActiveRow, String> {
    ensure_db_ready(app.clone(), db_state.inner()).await?;
    blocking::run("sort_mode_active_set", move || {
        sort_modes::set_active(&app, &cli_key, mode_id)
    })
    .await
}

#[tauri::command]
pub(crate) async fn sort_mode_providers_list(
    app: tauri::AppHandle,
    db_state: tauri::State<'_, DbInitState>,
    mode_id: i64,
    cli_key: String,
) -> Result<Vec<i64>, String> {
    ensure_db_ready(app.clone(), db_state.inner()).await?;
    blocking::run("sort_mode_providers_list", move || {
        sort_modes::list_mode_providers(&app, mode_id, &cli_key)
    })
    .await
}

#[tauri::command]
pub(crate) async fn sort_mode_providers_set_order(
    app: tauri::AppHandle,
    db_state: tauri::State<'_, DbInitState>,
    mode_id: i64,
    cli_key: String,
    ordered_provider_ids: Vec<i64>,
) -> Result<Vec<i64>, String> {
    ensure_db_ready(app.clone(), db_state.inner()).await?;
    blocking::run("sort_mode_providers_set_order", move || {
        sort_modes::set_mode_providers_order(&app, mode_id, &cli_key, ordered_provider_ids)
    })
    .await
}
